contracts/Rewarder.sol

```diff
@@ -0,0 +1,104 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.2;
+
+import "./Energy.sol";
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "hardhat/console.sol";
+
+contract Rewarder is Ownable {
+  Energy public energy;
+
+  struct Objective {
+    uint256 startingBlock;
+    uint256 endingBlock;
+    uint256 reward;
+    address origin;
+  }
+
+  Objective[] private objectives;
+
+  // origin address => user address => block list
+  mapping(address => mapping(address => uint256[])) public userInteractionsByOrigin;
+  // user address => objectiveIndex => claimed
+  mapping(address => mapping(uint256 => bool)) public objectivesCompletedByUser;
+
+  // un nombre mejor?
+  event UserInteraction(address indexed user, uint256 interactionIndex, address indexed originAddress);
+  event ObjectiveAdded(uint256 indexed index, uint256 reward, address indexed creator);
+  event RewardClaimed(address indexed user, uint256 reward, uint256 indexed objectiveIndex);
+
+  modifier hasObjective(address origin) {
+    bool objective = false;
+    for (uint256 i = 0; i < objectives.length; i++) {
+      if (objectives[i].origin == origin) {
+        objective = true;
+        break;
+      }
+    }
+    require(objective, "No reward assigned to msg.sender");
+    _;
+  }
+
+  constructor(address _energy) {
+    energy = Energy(_energy);
+  }
+
+  function getObjectives() external view returns (Objective[] memory) {
+    return objectives;
+  }
+
+  function onReward(address user) public hasObjective(msg.sender) returns (uint256 interactionIndex) {
```

Comment:
`hasObjective` has a problem, which is that is iterating through all objetives. If there are many this could cause very high gas usage and potentially be impossible to run because of the gas limit.

I don't think we need this check. If someone doesn't have an objetive available but still wants to call this they can. It even allows us to create retroactive objetives if we want.